integration_tests/src/main/python/data_gen.py

```diff
@@ -831,11 +832,15 @@ def _gen_scalars_common(data_gen, count, seed=0):
 
 def gen_scalars(data_gen, count, seed=0, force_no_nulls=False):
     """Generate scalar values."""
-    if force_no_nulls:
-        assert(not isinstance(data_gen, NullGen))
-    src = _gen_scalars_common(data_gen, count, seed=seed)
-    data_type = src.data_type
-    return (_mark_as_lit(src.gen(force_no_nulls=force_no_nulls), data_type) for i in range(0, count))
+    def gen_scalars_help(data_gen, count, seed, force_no_nulls):
+        if force_no_nulls:
+            assert(not isinstance(data_gen, NullGen))
+        src = _gen_scalars_common(data_gen, count, seed=seed)
+        data_type = src.data_type
+        return (_mark_as_lit(src.gen(force_no_nulls=force_no_nulls), data_type) for i in range(0, count))
+    return with_cpu_session(lambda spark: gen_scalars_help(data_gen=data_gen, 
```

Comment:
Putting this in a cpu_session fixes the current problem, but it adds a new one. If `gen_scalars` is called from inside a `with_*_session` it will have other problems. `with_spark_session` calls `reset_spark_session_conf` which does more then just reset the conf. It clears out the catalog too with no way to get the original config or catalog back after it exits.  That means with_gpu_session -> gen_scalars will result in the query running on the CPU after the gen_scalars.

I see a few ways to properly fix this.

1)  We set `spark.sql.legacy.allowNegativeScaleOfDecimal` when launching spark and have the test framework throw an exception if it is not set. Then we remove references to it in all of the tests for consistency. Then we file a follow on issue to fix `with_spark_session` to not allow nesting and to throw an exception if it is nested.
2 ) We fix with_spark_session to throw an exception if it is ever nested and do what you are doing today + update the docs for it to be clear that it can never be called from within a with_spark_session
3) We fix the test to call gen_scalar from within a with_spark_session and add a doc fix for gen_scalar to indicate that negative scale decimals can have problems if called from outside of with_spark_session block. Then we file a follow on issue to fix `with_spark_session` to not allow nesting and to throw an exception if it is nested.

I personally prefer option 1 but I am fine with option 2 or 3.  Talking to @jlowe he really prefers option 3. The main difference between option 3 and option 2 for me really about the amount of code that needs to change. If we just fix the one test and add some docs, that feels like a really small change. If we have to fix nesting/etc that feels a bit larger, but it is something we need to do either way and would mean all tests that use gen_scalar would be good to deal with all decimal values properly.

Reply:
I'm not a fan of 2.  It's again surprising behavior (who would expect it to spawn a Spark session?).  I'm fine with either 1 or 3, and even with 1, I still think we should fix the test(s).  We should be putting all data generation inside a spark session context of some kind.

Reply:
Thanks, updated code to option 3.
 
Now I warp all scalar generation with a `with_cpu_session`, no matter if it calls `f.lit` or uses `DecimalGen`. Not sure if we only want to move the cases that are possible to fail into Spark sessions.

Reply:
Follow-on issue: #9412